### PR TITLE
[FLINK-19326][cep] Allow explicitly configuring time behaviour on CEP PatternStream

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*With the deprecation of `StreamExecutionEnvironment.setStreamTimeCharacteristic()`, we need a way of explicitly configuring the time behaviour of CEP operations. Currently, all CEP operations use the characteristic to configure themselves. `PatternStream` could add `inProcessingTIme` and `inEventTime` that specify time behaviour to processing time or event time.*

## Brief change log

  - *`PatternStream` adds `inProcessingTIme` and `inEventTime` method instead of `setStreamTimeCharacteristic()` that specify time behaviour to processing time or event time.*

## Verifying this change

  - *Modify `CEPITCase` test cases which use `StreamExecutionEnvironment.setStreamTimeCharacteristic()` through `CEP.pattern(input, pattern, isProcessingTime`) to verify the `inProcessingTIme` of `PatternStream` whether could be instead of `setStreamTimeCharacteristic()`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@aljoscha Please help to review this pull request if you are available.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java [hunk 1]
```diff
@@ -60,14 +59,28 @@
 	 */
 	private final OutputTag<IN> lateDataOutputTag;
 
+	private boolean isProcessingTime;
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java:62]: I think it would be better to introduce an enum for this. Sth like `TimeCharacteristic` that has `PROCESSING_TIME` and `EVENT_TIME`. I find it's often hard to see what `true` or `false` mean when you just read a method call.

This should be internal, though, and the only API for users should be `inEventTime()` and `inProcessingTime()`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java [hunk 2]
```diff
@@ -52,7 +52,48 @@
...
+	 * @param input DataStream containing the input events
+	 * @param pattern Pattern specification which shall be detected
+	 * @param isProcessingTime Time behaviour to specify processing time or event time
+	 * @param <T> Type of the input events
+	 * @return Resulting pattern stream
+	 */
+	public static <T> PatternStream<T> pattern(
+			DataStream<T> input,
+			Pattern<T, ?> pattern,
+			boolean isProcessingTime) {
```
Comment (aljoscha) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java:70]: I think we don't need to expose an additional parameter because users can use `inProcessingTime()` and `inEventTime()` right?

Comment (aljoscha) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java:70]: It's better to just have one way of doing things. Keeps the API simpler. What do you think?

**Review by aljoscha — COMMENTED** [review]
I had a preliminary review. Overall this looks good but I had some comments on the API.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java (hunk 2 above)
Comment (SteNicholas) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java:70]: @aljoscha I thought that this could provide this interface to directly create PatternStream in processing time with parameter, and this is user-friendly interface. Your point that keeping the API simpler is right, but for users, IMO, it's better to specify the time behaviour by parameter, not extrally calling `inProcessingTime()` and `inEventTime()`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java (hunk 1 above)
Comment (SteNicholas) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java:62]: @aljoscha I previously consider the enum like `TimeCharacteristic`, but this enum is only used in this which makes little sense and confused for users in 1.11 version.

**Comment by SteNicholas** [comment]
> I had a preliminary review. Overall this looks good but I had some comments on the API.

@aljoscha thanks for your preliminary review. I have already replied the comments you mentioned. Please take a look if you are available.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java (hunk 1 above)
Comment (aljoscha) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java:62]: Users wouldn't see it because it's only internal. And having an enum will be more maintainable because, again, just seeing `true` or `false` in code makes it very hard to know what it means.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java (hunk 2 above)
Comment (aljoscha) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java:70]: I still think having this as a parameter would not be very user friendly, seeing a call like `pattern(in, foo, false)` makes it very hard to know what the `false` means. If you see, however, `pattern(in, foo).inProcessingTime()` it's immediately obvious what's going on.

Comment (SteNicholas) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/CEP.java:70]: @aljoscha I would like to follow up the points you mentioned later to update the methods.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java (hunk 1 above)
Comment (SteNicholas) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStreamBuilder.java:62]: @aljoscha Yeah. I would like to add enum for processing time and event time.

**Comment by SteNicholas** [comment]
@aljoscha ,thanks for your detailed explanation. Please help to review this pull requests again if you are available. Thanks.

**Comment by SteNicholas** [comment]
@aljoscha ,thanks for your detailed explanation. Please help to review this pull requests again if you are available. Thanks.

**Comment by SteNicholas** [comment]
@aljoscha ,thanks for your detailed explanation. Please review this pull requests again if you are available. Thanks very much.

**Comment by aljoscha** [comment]
@SteNicholas Thanks for your work! I pushed some changes to your branch, I hope that's ok. I split the unrelated changes that remove imports. I also fixed the `PatternStreamBuilder`, the problem with the code before was that the time characteristic would be dropped when calling the builder methods that don't call the right constructor. And I changed `DataStreamMatch` to use an explicit type instead of a boolean, because it's hard to parse what they mean when looking at the code in passing.

Please let me know if you're ok with the changed, then I'd merge once CI is green.

**Comment by SteNicholas** [comment]
> Thanks for your work! I pushed some changes to your branch, I hope that's ok. I split the unrelated changes that remove imports. I also fixed the `PatternStreamBuilder`, the problem with the code before was that the time characteristic would be dropped when calling the builder methods that don't call the right constructor. And I changed `DataStreamMatch` to use an explicit type instead of a boolean, because it's hard to parse what they mean when looking at the code in passing.
> 
> Please let me know if you're ok with the changed, then I'd merge once CI is green.

**Comment by aljoscha** [comment]
@flinkbot run azure

**Comment by SteNicholas** [comment]
> @SteNicholas Thanks for your work! I pushed some changes to your branch, I hope that's ok. I split the unrelated changes that remove imports. I also fixed the `PatternStreamBuilder`, the problem with the code before was that the time characteristic would be dropped when calling the builder methods that don't call the right constructor. And I changed `DataStreamMatch` to use an explicit type instead of a boolean, because it's hard to parse what they mean when looking at the code in passing.
> 
> Please let me know if you're ok with the changed, then I'd merge once CI is green.

@aljoscha Your changes are ok for me. Please go ahead to do some changes. And no any comments with my changes?

**Comment by aljoscha** [comment]
Ah sorry, your other changes took very good. The API is very clean now. 👌

**Comment by SteNicholas** [comment]
> Ah sorry, your other changes took very good. The API is very clean now. 👌

@aljoscha , thanks for your detailed review.

**Comment by aljoscha** [comment]
Good work, I merged this!
